### ISSUE TITLE
Load automatic fixes from allele_qc

### ISSUE THREAD

**Issue by manulera** [pombase/canto#2689]

**Comment by kimrutherford** [comment]
Thanks Manu.  I've transfer the issue to the Canto repo.

**Comment by manulera** [comment]
Related to https://github.com/pombase/pombase-chado/issues/1021

**Comment by manulera** [comment]
## Manual changes

@kimrutherford

After what we discussed today, I prepared the following files in the allele_qc repo:

```
manual_fixes_pombase/
├── manual_alleles2remove.tsv
└── manual_changes_formatted.tsv
```

Columns are self-explaining

These are the manual changes so in the future tsv files with the same structure should be usable instead to add bulk changes using the same script you make for this.

**Comment by manulera** [comment]
## Automated changes

These are in:

```
results/
├── allele_auto_fix.tsv
├── protein_modification_auto_fix.tsv
```

For alleles, the relevant columns are:
* systematic_id, allele_description, allele_name, allele_type: to constitute unique identifier
* change_description_to, change_type_to, auto_fix_comment: fixes + comments
* solution_index: to accept the fix, this column must be empty. It was values 0-N where N is the number of possible solutions that have been found. These must be resolved manually.
* rest of columns: extra info from the analysis.

For protein modifications, I was not sure what to remove and what to keep, so I kept all columns, because there are some with the extension, etc. The new columns that can be used for fixing:
* sequence_error: indicates the residues that do not match or fall outside of the protein sequence. Can be used to double-check.
* change_sequence_position_to, auto_fix_comment: fix and a comment to explain it
* solution_index: same as above

**Comment by manulera** [comment]
## Known errors

```
results/
├── allele_cannot_fix_other_errors.tsv
├── allele_cannot_fix_sequence_errors.tsv
├── protein_modification_cannot_fix_other_errors.tsv
├── protein_modification_cannot_fix_sequence_errors.tsv
```

Sequence errors have the unique identifier columns (all columns for protein mods) + sequence_error column with the residues that do not match / positions outside of the chain length.

For "other errors":
* protein modifications: extra category "error" with self-explaining categorical values
* alleles: error_type, error_info pairs, since errors can be a bit more complicated

**Comment by manulera** [comment]
Once we fix this, I can add the info above to the repo documentation.

**Comment by kimrutherford** [comment]
Hi @manulera 

I'm working on this today.  Could you help me with the two changes below?  I don't know if this an error in the file or I don't understand, but the same allele name/description/type appears twice with a different "change_description_to" value.  Which change should be applied in this case?

| systematic_id | allele_description | allele_name | allele_type         | change_description_to | change_type_to | auto_fix_comment | sequence_error | solution_index | allele_parts | rules_applied                 | reference    |
| ------------- | ------------------ | ----------- | ------------------- | --------------------- | -------------- | ---------------- | -------------- | -------------- | ------------ | ----------------------------- | ------------ |
| SPAC3A12.14   | E104V              | cam1-E124   | amino_acid_mutation | E86V                  |                | multi_shift_fix  | E104           | 0              | E104V        | amino_acid_mutation:single_aa | PMID:7657644 |
| SPAC3A12.14   | E104V              | cam1-E124   | amino_acid_mutation | E106V                 |                | multi_shift_fix  | E104           | 1              | E104V        | amino_acid_mutation:single_aa | PMID:7657644 |

**Comment by kimrutherford** [comment]
>  Could you help me with the two changes below?

Please ignore that.  :-)

I've now re-read your comments and I see that you've documented these cases:

> solution_index: to accept the fix, this column must be empty. It was values 0-N where N is the number of possible solutions that have been found. These must be resolved manually.

**Comment by kimrutherford** [comment]
In the file there are four cases where we should change the type to `amino_acid_other`.  That isn't configured in Canto yet.  Should we add it to Canto or just use allele type `other`?

**Comment by kimrutherford** [comment]
In the case below we will change description "K56R"  to "K57R".
  - should we change "hht2-K56R" to "hht2-K57R"?
  - if we change the name should "hht2-K56R" be kept as a synonym?

(Sorry if we talked about this and I forgot)

| systematic_id | allele_description | allele_name | allele_type         | change_description_to | change_type_to | auto_fix_comment | sequence_error | solution_index | allele_parts | rules_applied                 | reference                   |
| ------------- | ------------------ | ----------- | ------------------- | --------------------- | -------------- | ---------------- | -------------- | -------------- | ------------ | ----------------------------- | --------------------------- |
| SPBC8D2.04    | K56R               | hht2-K56R   | amino_acid_mutation | K57R                  |                | histone_fix      | K56            |                | K56R         | amino_acid_mutation:single_aa | PMID:17369611,PMID:34608864 |

**Comment by kimrutherford** [comment]
I've made a first attempt [at a script](https://github.com/pombase/canto/blob/master/etc/apply_allele_fixes.pl) for processing `allele_auto_fix.tsv` and `manual_changes_formatted.tsv`.  I done a bit of testing on my desktop to make sure it's all OK but I'll do more before applying it to the main Canto.

The only things the isn't handled yet are the `chimera` "alleles".  Maybe we could have a quick chant about those on Tuesday?  I want to double check that I know what I should be doing with those.

**Comment by kimrutherford** [comment]
After fixing Canto we'll need to fix the PHAF files.  Some have quite a few non-deletion alleles.  eg.
`pombe-embl/external_data/phaf_files/chado_load/htp_phafs/PMID_24362309_phaf.tsv`

I did a quick grep for allele names from `allele_auto_fix.tsv` and found 192 PHAF file rows that mention them.

**Comment by manulera** [comment]
> In the file there are four cases where we should change the type to amino_acid_other. That isn't configured in Canto yet. Should we add it to Canto or just use allele type other?

I have not discussed this with @ValWood yet. I made an issue about it (https://github.com/pombase/allele_qc/issues/17), but never brought it up. I think we should have this type for alleles with mutations described at the aminoacid level that do not fall into the existing categories, because they have new combinations. For example amino_acid_mutation + nonsense_mutation or nonsense_mutation + amino_acid_insertion. We should use amino_acid_other over other, to be specific that the changes are described at the protein level.

**Comment by manulera** [comment]
> In the case below we will change description "K56R" to "K57R".

Yes, this should be done and perhaps it's easiest to do it at the same time. Let's discuss it tomorrow as well. It should be fairly simple to identify those that have the wrong description in the name.
 
> if we change the name should "hht2-K56R" be kept as a synonym?
Yes, we whould

**Comment by manulera** [comment]
> After fixing Canto we'll need to fix the PHAF files. Some have quite a few non-deletion alleles. eg.

I can run it on those as well.

**Comment by manulera** [comment]
So things to discuss tomorrow:
* Chimeras
* Change of names
* aminoacid_other

**Comment by manulera** [comment]
Yes, that's right @ValWood, I did not realise that example was a histone. We know from the comments whether the reason for the fix was shifting the coordinates of a histone, so I can filter those out.

**Comment by ValWood** [comment]
So, today we decided that changing an amino acid to a stop codon should be a type of 
amino acid deletion. Maybe we should rename the class to make it clearer  "deletion or truncation"?  (but would also need to do for the compound terms and they may get a bit long)?

**Comment by ValWood** [comment]
How many  "nonsense mutation" do we have out of curiosity?
If we do capture "nonsense mutation" with amino acid deletions, we should still allow these to. be described as  S123->stop (or ochre/amber/ the other one)

**Comment by manulera** [comment]
> How many "nonsense mutation" do we have out of curiosity?

100

> If we do capture "nonsense mutation" with amino acid deletions, we should still allow these to. be described as S123->stop (or ochre/amber/ the other one)

Yes, we should (current syntax is `S123*`)

> So, today we decided that changing an amino acid to a stop codon should be a type of
amino acid deletion

I added that to the pipeline

> Maybe we should rename the class to make it clearer "deletion or truncation"? (but would also need to do for the compound terms and they may get a bit long)?

I think it's OK. If we remove the option to have the nonsense mutation, people will probably check the aminoacid_deletion. For people that have used Canto before, we can send an announcement in the mailing list. I have modified the examples given by Canto so that they show the nonsense mutation syntax. @kimrutherford this PR: https://github.com/pombase/canto/pull/2711

At some point we should remove the option from Canto, but I leave that for you @kimrutherford

**Comment by manulera** [comment]
@kimrutherford I have added the extra column `change_name_to` to the auto-fix file https://github.com/pombase/allele_qc/blob/master/results/allele_auto_fix.tsv

It checks whether the old description was in the name, and if so it replaces it by the new description. It's not a full-proof solution, as it may miss some, but I inspected the list and I think it caught all the errors.

As we said, it excludes the histone fixes.

**Comment by kimrutherford** [comment]
> I have added the extra column change_name_to to the auto-fix file

Thanks!  I've changed my script to use it correctly.  Should the old allele name become a synonym?

**Comment by ValWood** [comment]
> Should the old allele name become a synonym?

yes please. This is how it would be in the publication. So ev entually, when we have allele pages, and can search alleles we wany people to be able to find with the original names

**Comment by kimrutherford** [comment]
> The only things the isn't handled yet are the chimera "alleles".

That's fixed now so that the script will store chimeras as type `fusion_or_chimera`.

> After fixing Canto we'll need to fix the PHAF files.

I've added a script to do that.  It changes 191 PHAF lines.  Here is the output from script (the list of changes applied) for all the PHAF files: [phaf_allele_fix_log.txt](https://github.com/pombase/canto/files/11045522/phaf_allele_fix_log.txt)

The only thing I'd like to check with you both are the changes for `rad16-U` in the PHAF files.  The script currently changes the type of `rad16-U` alleles from `unknown` to `nonsense_mutation` and the description from `unknown` to `Q643*`.  Are we sure that there aren't multiple alleles with the name `rad16-U`?  It seems like a short and uninformative name.  Does the "U" in the name stand for "Unknown"?

An example file with that allele is: `external_data/phaf_files/chado_load/ltp_phafs/PMID_9658208_phaf.tsv`

Apart from that question, it all looks OK.

**Comment by kimrutherford** [comment]
> I've made a first attempt [at a script](https://github.com/pombase/canto/blob/master/etc/apply_allele_fixes.pl) for processing allele_auto_fix.tsv and manual_changes_formatted.tsv. I done a bit of testing on my desktop to make sure it's all OK but I'll do more before applying it to the main Canto.

One problem I forgot about is that there are still a lot of alleles in Canto that don't have a gene name prefix (like "D226N").  We probably should have fixed those first.

The allele names in Manu's file have the corrected names with the gene name prefix (like "cce1-D226N").  The allele fixing script was missing a bunch of alleles because it was looking for the prefixed names, which weren't always there in Canto.  I've changed the allele fixing script to handle that now.

**Comment by kimrutherford** [comment]
> The allele fixing script was missing a bunch of alleles

Turns out that "a bunch" was actually 13.  :-)

**Comment by kimrutherford** [comment]
Hi @manulera

There are a few lines in `manual_changes_formatted.tsv` that are duplicates.  Do you think that's a problem?

These lines appear twice:

| systematic_id | allele_name | allele_description                                                | change_description_to                                             | change_name_to | change_type_to | comment |
| ------------- | ----------- | ----------------------------------------------------------------- | ----------------------------------------------------------------- | -------------- | -------------- | ------- |
| SPAC24H6.06   | sld3-9A     | S140A,T201A,T228A,S449A,T636A,T650A,S673A,T690A,S698A             | S140A,T201A,T228A,S499A,T636A,T650A,S673A,T690A,S698A             |                |                |         |
| SPAC24H6.06   | sld3-N4A    | S140A,T201A,T228A,S449A                                           | S140A,T201A,T228A,S499A                                           |                |                |         |
| SPBC11C11.02  | imp2-11A    | S346A,S367A,T384A,S397A,S402A,T426A,S452A,S467A,S490A,S511A,T531A | T346A,S367A,T384A,S397A,S402A,T426A,S451A,T467A,S490A,S511A,T531A |                |                |         |
| SPBC11C11.02  | imp2-11E    | S346E,S367E,T384E,S397E,S402E,T426E,S452E,S467E,S490E,S511E,T531E | T346E,S367E,T384E,S397E,S402E,T426E,S451E,T467E,S490E,S511E,T531E |                |                |         |
| SPBC32F12.04  | gtb1-K5A    | K399A                                                             | K389A                                                             |                |                |         |

**Comment by manulera** [comment]
I don't know why that's the case, but if they both say the same I would say they are probably correct.

**Comment by manulera** [comment]
> rad16-U

The fix comes from this publication, where the change to STOP was curated as Q643->amber. Maybe in the HTP paper they didn't bother to write the modification of mutants? Or maybe this was sequenced a posteriori?

https://www.pombase.org/reference/PMID:9658208

**Comment by kimrutherford** [comment]
While testing my script reading `allele_qc/manual_fixes_pombase/manual_changes_formatted.tsv` I get the following warnings.  They all look harmless but I thought I should report them just in case:

```
for "kin1-TA" description in DB doesn't match file: "T299A" vs "T2999A"
for "myo2-ST" description in DB doesn't match file: "chimera+myp2-aa1923-2104" vs "+myp2-aa1923-2104"
for "myo2-LT" description in DB doesn't match file: "chimera + myp2-aa1622-2104" vs "+ myp2-aa1622-2104"
```

**Comment by manulera** [comment]
Good catch, I fixed the myo2 ones and other chimeras:

https://github.com/pombase/allele_qc/commit/25d67db8208642fa5eebc2f451696cfcbebb37fc

**Comment by kimrutherford** [comment]
Files to do (for kmr):

 - `allele_auto_fix.tsv`
 - `manual_changes_formatted.tsv`
 - `manual_fix_chimeras.tsv`
    - process `add_comment` column

**Comment by kimrutherford** [comment]
> process add_comment column

Also process and store the `add_synonym` column from `manual_fix_chimeras.tsv`

**Comment by kimrutherford** [comment]
I didn't finish this issue.  I got distracted.

It's all looking OK I think.  I've committed all the PHAF file fixes.  I'm doing some final testing of the Canto fixes and I'll apply those changes on Wednesday night UK time.

**Comment by manulera** [comment]
To the manual ones

**Comment by kimrutherford** [comment]
> I'm doing some final testing of the Canto fixes and I'll apply those changes on Wednesday night UK time.

I forgot but I'm doing that now.  Canto will be offline for an hour or so.

**Comment by kimrutherford** [comment]
All the allele fixes have been applied to Canto and to the PHAF files.  I'll have a good look at the results of the Chado load on Tuesday morning to make sure it's all OK.

> I keep adding lines to the files, so give me a heads-up to move those changes to the changes history once they have been ingested.

It's up to date now.   The fixes from these file have been applied:

  - allele_auto_fix.tsv
  - manual_changes_formatted.tsv
  - manual_fix_chimeras.tsv

**Comment by kimrutherford** [comment]
> All the allele fixes have been applied to Canto and to the PHAF files.

But then I forgot to commit all the PHAF file fixes to SVN.  Whoops.

That's done now.

**Comment by ValWood** [comment]
Yipee alleles of type "other" down to
(209 rows)

**Comment by manulera** [comment]
Ok, I have added these to the changelog, with the date in this commit: https://github.com/pombase/allele_qc/commit/7f09eb875df6499a55c655cd3ca887002fedca2c

So that's good news! 🥳

But:

* I forgot to include these in the previous one: https://github.com/pombase/allele_qc/blob/master/manual_fixes_pombase/allele_manual_changes_formatted.tsv
* We already have 34 new errors: https://github.com/pombase/allele_qc/blob/master/results/allele_auto_fix.tsv
    * This is excluding the cam1 alleles that I still have to manually fix (the ones that have a value in the `solution_index column`)

We can anyway close this issue, and I will make a new one when I have fixed the manual ones.
